### PR TITLE
Refactor/email confirm links

### PR DESCRIPTION
**Pull request by zomars**
## What does this PR do?

- Reverts some changes made in #5939
- Moves confirmation logic to `/api/link`
- Reuses Booking UI to display status
- Drops adding rejecting reason support (for now)
- Reuses tRPC logic
- Uses encryption/decryption in a single token instead of hashing

To do (maybe)

- Add backwards compatibility with old token (IDK if worth it)

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Chore (refactoring code, technical debt, workflow improvements)

## How should this be tested?

- [ ] Book a event with "requires confirmation"
- [ ] Test Accept/Reject buttons

### CODE REVIEW

**Review by zomars — COMMENTED** [review]
Ready for review

File: apps/web/playwright/webhook.e2e.ts [hunk 1]
```diff
@@ -57,6 +57,7 @@ test("add webhook & test that creating an event triggers a webhook call", async
     attendee.timeZone = dynamic;
     attendee.language = dynamic;
   }
+  body.payload.organizer.id = dynamic;
```
Comment (zomars) [inline apps/web/playwright/webhook.e2e.ts:60]: We now need this organizer id to generate the token

File: packages/lib/constants.ts [hunk 2]
```diff
@@ -53,3 +53,4 @@ export const IS_STRIPE_ENABLED = !!(
 );
 /** Self hosted shouldn't checkout when creating teams unless required */
 export const IS_TEAM_BILLING_ENABLED = IS_STRIPE_ENABLED && (!IS_SELF_HOSTED || HOSTED_CAL_FEATURES);
+export const CALENDSO_ENCRYPTION_KEY = process.env.CALENDSO_ENCRYPTION_KEY || "";
```
Comment (emrysal) [inline packages/lib/constants.ts:56]: We could be easily leaking the CALENDSO_ENCRYPTION_KEY here, if there is any client side code including packages/lib/constants.ts - also this hardcodes it into the build (that's not that much of an issue and may be desirable)

**Review by emrysal — APPROVED** [review]
It looks good to me, nice cleanup, have a look at my comments and see if there's any merrit to my thought - will turn off automerge so you can eval.

**Review by alannnc — APPROVED** [review]
It works and couldn't find any security leak.

File: apps/web/pages/api/link.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+import { z } from "zod";
+
+import { CALENDSO_ENCRYPTION_KEY } from "@calcom/lib/constants";
+import { symmetricDecrypt } from "@calcom/lib/crypto";
+import { defaultResponder } from "@calcom/lib/server";
+import prisma from "@calcom/prisma";
+import { createContext } from "@calcom/trpc/server/createContext";
+import { viewerRouter } from "@calcom/trpc/server/routers/viewer";
+
+enum DirectAction {
```
[the remark this answers is not in the document]
Comment (zomars) [reply apps/web/pages/api/link.ts:12]: just kept it as @leog made it. No particular reason

File: packages/lib/constants.ts (hunk 2 above)
Comment (zomars) [reply packages/lib/constants.ts:56]: Good point

Comment (zomars) [reply packages/lib/constants.ts:56]: We use this pattern in multiple places tho 
image

Comment (zomars) [reply packages/lib/constants.ts]: Removed for now

Comment (emrysal) [reply packages/lib/constants.ts]: Those are OK - all those are serverside 👍  (😌)

**Review by leog — APPROVED** [review]
LGTM. Found some NITs comparing to old approach:
1. When an error happens (invalid token, booking already confirmed or rejected), we should redirect to page 500 with an error query param with the message
2. I think we should remove the following section of the booking page for this feature
<kbd>image</kbd>

File: apps/web/pages/api/link.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (zomars) [reply apps/web/pages/api/link.ts:11]: image

updated

**Comment by zomars** [comment]
> LGTM. Found some NITs comparing to old approach:
> 
> 1. When an error happens (invalid token, booking already confirmed or rejected), we should redirect to page 500 with an error query param with the message

Agreed, we can follow up tho. Not critical IMO.

> 2. I think we should remove the following section of the booking page for this feature
>    image

We can discuss this. I don't think is that much of an issue.

**Comment by leog** [comment]
> Agreed, we can follow up tho. Not critical IMO.

Agreed, which is why I approved it anyway. NITs shouldn't stop a PR.

> We can discuss this. I don't think is that much of an issue.

Not an issue, it just looks weird, an oversight. The person accepting/rejecting already has an account, it just looks sloppy IMO.

**Pull request merged**